handler/artifact_upload_handler.py

```diff
@@ -0,0 +1,115 @@
+"""
+A tornado application to handle uploading a conda build artifact to a specified
+file location.
+
+"""
+
+import argparse
+import hashlib
+import hmac
+import os
+import subprocess
+
+import tornado.ioloop
+import tornado.httpserver
+import tornado.web
+
+
+class ArtifactUploadHandler(tornado.web.RequestHandler):
+    """Handle writing conda build artifacts to a channel."""
+
+    def initialize(self, write_path, conda_exe, hash_expected):
+        """
+        "Hook for subclass initialization."
+
+        In this class, when we initialize `ArtifactUploadHandler`, we need to
+        set:
+         * the path that incoming files will be written to,
+         * the path to the conda executable to use for indexing the channel,
+         * the expected hashed and salted digest of the secure upload token.
+
+        """
+        self.write_path = write_path
+        self.conda_exe = conda_exe
+        self.hash_expected = hash_expected
+
+    def _check_token(self, token):
+        """
+        Check we have been passed the *correct* secure token before uploading
+        the artifact to the channel.
+        
+        """
+        salt = '42679ad04d44c96ed27470c02bfb28c3'
+        to_hash = '{}{}'.format(salt, token)
+        hash_result = hashlib.sha256(to_hash).hexdigest()
+        # Reduce the risk of timing analysis attacks.
+        return hmac.compare_digest(self.hash_expected, hash_result)
+
+    def _index_channel(self):
+        """
+        Index the conda channel at `self.write_path` to ensure the channel
+        remains consistent.
+        
+        """
+        cmds = [self.conda_exe, 'index', self.write_path]
+        subprocess.check_call(cmds)
+
+    def post(self):
+        token = self.get_argument('token')
+        file_data, = self.request.files['artifact']
+        filename = file_data['filename']
+        body = file_data['body']
+        if self._check_token(token):
+            with open(os.path.join(self.write_path, filename), 'w') as owfh:
+                owfh.write(body)
+            self._index_channel()
+
+
+def make_app(write_path, conda_exe):
+    kw = {'write_path' :write_path,
```

Comment:
@dkillick Minor nitpick, but the spacing is out i.e. should be `'write_path': write_path,'`

Reply:
@dkillick BTW Is this file PEP8 or flake8 compliant? ... just askin'

Reply:
Argh, that colon should not be there! Given that it is, this file is very likely not flake8 compliant!

Reply:
Boom. `flake8` compliant (at next commit) 💥